### PR TITLE
Add support for compiling arbitrary unitaries to `Chip0`

### PR DESCRIPTION
**Pull request by josh146**
**Description of the Change:**

* Adds the `MZgate` to `ops.py`, representing a Mach-Zehnder interferometer. This is not a primitive of the existing simulator backends; rather, `_decompose()` is defined, decomposing it into an external phase shift, two 50-50 beamsplitters, and an internal phase shift.

* Several changes to the `sf.decompositions` module:
  
  - The name `clements` -> `rectangular`
  - All interferometer decompositions (`rectangular`, `rectangular_phase_end`, `rectangular_symmetric`, and `triangular`) now have standardized outputs `(tlist, diag, tilist)`, so they can easily be swapped.

* Several changes to `ops.Interferometer`:

  - The calculation of the `ops.Interferometer` decomposition has been moved from `__init__` to `_decompose()`, allowing the interferometer decomposition type to be set by a `CircuitSpec` during compilation.

  - `**kwargs` is now passed through from `Operation.decompose` -> `Gate.decompose` -> `SpecificOp._decompose`, allowing decomposition options to be passed during compilation.

  - `ops.Interferometer` now accepts the keyword argument `mesh` to be set during initialization, allowing the user to specify the decomposition they want.

* Moves the `compile_seq` function from `Program.compile` to a `CircuitSpecs` method. This allows it to be accessed from the `CircuitSpec.compile` method. Furthermore, it now must also be passed the program registers, as compilation may sometimes require this.

* Allow `GraphEmbed` to pass kwargs to `Interferometer`.

* Modify `Chip0Spec` such that it groups operations, ensures the first set is `S2gate`s, the second set is `{BSgate, Rgate}`, and the third set is `MeasureFock()`

* Combine the third set into a single `MeasureFock` call, as in `GBSSpec`

* Combine the second set int a single unitary via matrix multiplication, and decompose via `mesh="rectangular_symmetric"`.

**Benefits:**

* Any arbitrary unitary, consisting of `{BSgate, Rgate, Interferometer, MZgate}` gates, placed between `S2gate`s and `MeasureFock`, can now be compiled to run on Chip0.

* `CircuitSpec`s can now pass decomposition options to quantum operations `_decompose()` method.

* Users now have access to alternate decompositions for unitaries; previously, only `rectangular` was supported directly from the SF frontend

**Possible Drawbacks:**

* Since unitary decomposition has been moved from `Interferometer.__init__` to `Interferometer._decompose`, any errors that are raised during decomposition will no longer be raised during program construction, but during `eng.run()`. Unfortunately, this is unavoidable, if we wish to make the decomposition type a compilation option.

**Related GitHub Issues:** #115, #103

### CODE REVIEW

File: strawberryfields/ops.py [hunk 1]
```diff
@@ -1361,6 +1362,30 @@ def _apply(self, reg, backend, **kwargs):
...
 
+class MZgate(Gate):
+    r"""Mach-Zehnder interferometer.
+
+    .. math::
+        MZ(\phi_1, \phi_2) = BS(\pi/4, \pi/2) (R(\phi_1)\otimes I) BS(\pi/4, \pi/2) (R(\phi_2)\otimes I)
+
+    Args:
+        phi1 (float): internal phase
+        phi2 (float): external phase
```
Comment (smite) [inline strawberryfields/ops.py]: Maybe consider swapping the names phi1 and phi2? Temporally phi2 takes place first.

File: strawberryfields/ops.py [hunk 2]
```diff
@@ -1615,46 +1640,110 @@ def __or__(self, reg):
...
+    def __init__(self, U, mesh="rectangular", tol=1e-6):
         super().__init__([U])
         self.ns = U.shape[0]
+        self.mesh = mesh
+        self.tol = tol
+
+        allowed_meshes = {"rectangular", "rectangular_phase_end", "rectangular_symmetric", "triangular"}
+
+        if mesh not in allowed_meshes:
+            raise ValueError("Unknown mesh {}. Mesh must be one of {}".format(mesh, allowed_meshes))
```
Comment (smite) [inline strawberryfields/ops.py]:
```suggestion
            raise ValueError("Unknown mesh '{}'. Mesh must be one of {}".format(mesh, allowed_meshes))
```

File: strawberryfields/circuitspecs/chip0.py [hunk 3]
```diff
@@ -51,7 +56,42 @@ class Chip0Specs(CircuitSpecs):
...
+
         # Measurement in Fock basis
         MeasureFock() | [0, 1, 2, 3]
         """
     )
+
+    def compile(self, seq):
+        # first do general GBS compilation to make sure
+        # Fock measurements are correct
+        seq = GBSSpecs().compile(seq)
```
Comment (smite) [inline strawberryfields/circuitspecs/chip0.py]: Another option: inherit GBSSpec, call `super().compile()` here?

Comment (josh146) [reply strawberryfields/circuitspecs/chip0.py]: This makes it difficult to call `Circuit Spec.compile` for the final topology check, though

File: strawberryfields/ops.py (hunk 1 above)
Comment (josh146) [reply strawberryfields/ops.py]: I did this in order to match the order provided by the `rectangular_symmetric` decomposition - if it makes more sense, we should change both.

File: tests/frontend/test_circuitspecs_chip0.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+            # corresponds to an identity on modes [2, 3]
+            ops.Rgate(0) | q[2]
+            ops.BSgate(np.pi/4, np.pi/2) | (q[2], q[3])
+            ops.Rgate(np.pi) | q[2]
+            ops.BSgate(np.pi/4, np.pi/2) | (q[2], q[3])
+            ops.Rgate(np.pi) | q[2]
+            ops.Rgate(0) | q[3]
+
+            ops.MeasureFock() | (q[0], q[3], q[1], q[2])
```
Comment (josh146) [inline tests/frontend/test_circuitspecs_chip0.py]: Note: `GBSSpec.compile()` seems to permute the measured mode order. I don't _think_ this is an issue anywhere, but may be unintuitive to the user, and worth fixing in another PR.

File: tests/frontend/test_circuitspecs_chip0.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+            phi3=0,
+        )
+
+        expected = to_program(bb)
+        res = expected.compile("chip0")
+
+        for cmd1, cmd2 in zip(res.circuit, expected.circuit):
+            assert cmd1.op.__class__ == cmd2.op.__class__
+            assert [p.x for p in cmd1.op.p] == [p.x for p in cmd2.op.p]
+            assert [i.ind for i in cmd1.reg] == [i.ind for i in cmd2.reg]
```
Comment (josh146) [inline tests/frontend/test_circuitspecs_chip0.py]: This is a really cumbersome way of checking the equality of two programs. A useful feature would be to overload the `Program.__eq__(self, other)` special method, in order to perform the `nx.is_isomorphic` check (similarly to that done in `CircuitSpecs.compile`, but also taking into account parameter values).

That would allow us to simply do `assert prog1 == prog2` in the tests.

File: tests/frontend/test_circuitspecs_chip0.py (hunk 4 above)
Comment (co9olguy) [reply tests/frontend/test_circuitspecs_chip0.py]: I seem to recall TFBackend gives an exception if modes are out of order in `measure_fock`

Comment (josh146) [reply tests/frontend/test_circuitspecs_chip0.py]: Ah, so worthwhile pinning down and fixing then. I'm not really sure where the permutation happens, but my best guess is `GBSSpec.compile`

Comment (co9olguy) [reply tests/frontend/test_circuitspecs_chip0.py]: seems likely it's happening within one of the calls to networkx

File: strawberryfields/ops.py (hunk 1 above)
Comment (josh146) [reply strawberryfields/ops.py]: Done

File: strawberryfields/circuitspecs/chip0.py (hunk 3 above)
Comment (co9olguy) [reply strawberryfields/circuitspecs/chip0.py]: @josh146 how come?

File: strawberryfields/circuitspecs/chip0.py [hunk 6]
```diff
@@ -51,7 +59,130 @@ class Chip0Specs(CircuitSpecs):
         Rgate({internal_phase_0}) | [2]
         BSgate(pi/4, pi/2) | [2, 3]
 
+        # final local phases
+        Rgate({phi0}) | 0
```
[the remark this answers is not in the document]
Comment (co9olguy) [reply strawberryfields/circuitspecs/chip0.py]: Might be good for @lneuhaus to verify this modified circuit specification

File: strawberryfields/circuitspecs/chip0.py (hunk 3 above)
Comment (josh146) [reply strawberryfields/circuitspecs/chip0.py]: Because Python doesn't really provide a way to call methods from the grandfather class. The recommended approach is to invoke the grandfather class directly, but this doesn't work in this case, since `CircuitSpec` is an abstract base class and cannot be instantiated.

File: strawberryfields/circuitspecs/gaussian.py [hunk 7]
```diff
@@ -62,4 +62,5 @@ class GaussianSpecs(CircuitSpecs):
         "S2gate": {},
         "CXgate": {},
         "CZgate": {},
+        "MZgate": {},
```
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/circuitspecs/gaussian.py]: I don't know :laughing: I think @lneuhaus suggested `MZgate` during a chat, and it was the first thing in my head. Plus, less letters to type?

File: tests/frontend/test_ops_decompositions.py [hunk 8]
```diff
@@ -165,7 +165,7 @@ def test_decomposition(self, tol):
         n = 3
         prog = sf.Program(n)
         U = random_interferometer(n)
-        BS1, BS2, R = dec.clements(U)
+        BS1, R, BS2 = dec.rectangular(U)
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/frontend/test_ops_decompositions.py]: Oh, strange. Should I fix this here, or create a new PR?

File: tests/frontend/test_circuitspecs_chip0.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+        expected = sf.Program(4)
+
+        with expected.context as q:
+            ops.S2gate(0.5, 0) | (q[0], q[2])
+            ops.S2gate(0.5, 0) | (q[1], q[3])
+
+            # corresponds to MZgate(phi, theta) on modes [0, 1]
+            ops.Rgate(phi) | q[0]
+            ops.BSgate(np.pi/4, np.pi/2) | (q[0], q[1])
+            ops.Rgate(theta+2*np.pi) | q[0]
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/frontend/test_circuitspecs_chip0.py]: Yep, the issue is the `sf.decompositions.rectangular_decomposition` function. The best approach is to modify it so that the output decomposition parameters are always `phi % 2*np.pi`, or something similar, to force it within the range `[0, 2*pi)`.

@lneuhaus, what range of angle parameters do you prefer on your end, to minimize the number of conversions we have to do?

File: tests/frontend/test_circuitspecs_chip0.py [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+            ops.BSgate(np.pi/4, np.pi/2) | (q[2], q[3])
+            ops.Rgate(3*np.pi/4) | (q[2])
+            ops.Rgate(-np.pi/4) | (q[3])
+
+            ops.MeasureFock() | (q[0], q[3], q[1], q[2])
+
+        for cmd1, cmd2 in zip(res.circuit, expected.circuit):
+            assert cmd1.op.__class__ == cmd2.op.__class__
+            assert [p.x for p in cmd1.op.p] == [p.x for p in cmd2.op.p]
+            assert [i.ind for i in cmd1.reg] == [i.ind for i in cmd2.reg]
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/frontend/test_circuitspecs_chip0.py]: I ran the `rectangular_decomposition` function, looked at the output (it was easy to see the fraction of pi the parameters corresponded to), and then manually did the matrix multiplication by hand to verify it was correct :laughing:

File: strawberryfields/ops.py [hunk 11]
```diff
@@ -1361,6 +1362,31 @@ def _apply(self, reg, backend, **kwargs):
...
 
 
+class MZgate(Gate):
+    r"""Mach-Zehnder interferometer.
+
+    .. math::
+        MZ(\phi_1, \phi_2) = BS(\pi/4, \pi/2) (R(\phi_2)\otimes I) BS(\pi/4, \pi/2) (R(\phi_1)\otimes I)
+
+    Args:
+        phi1 (float): external phase
```
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/ops.py]: That is better, will change :+1:

File: strawberryfields/ops.py [hunk 12]
```diff
@@ -1615,46 +1641,120 @@ def __or__(self, reg):
...
+
+                else:
+                    # Clements style beamsplitters
+                    if not (drop_identity and phi == 0):
+                        cmds.append(Command(Rgate(phi), reg[n]))
+
+                    if not (drop_identity and theta == 0):
+                        cmds.append(Command(BSgate(theta, 0), (reg[n], reg[m])))
+
+            for n, expphi in enumerate(R):
```
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/ops.py]: Probably, unless @nquesada coded it this way for something we are missing.

Although, I am a bit hesitant to make the change in this PR, as it seems out of scope, and would require changes in all the decomposition tests.

File: strawberryfields/ops.py [hunk 13]
```diff
@@ -1615,46 +1641,120 @@ def __or__(self, reg):
...
-                    cmds.append(Command(Rgate(-phi), reg[n]))
+            if BS2 is not None:
+                # Clements style beamsplitters
+
+                for n, m, theta, phi, _ in reversed(BS2):
+                    theta = theta if np.abs(theta) >= _decomposition_tol else 0
+                    phi = phi if np.abs(phi) >= _decomposition_tol else 0
+
+                    if not (drop_identity and theta == 0):
+                        cmds.append(Command(BSgate(-theta, 0), (reg[n], reg[m])))
```
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/ops.py]: Yep. The `sf.decompositions.rectangular` function returns the first array of beamsplitters, local phase shifts, and then the final array of inverse beamsplitters in reversed order (I think this corresponds exactly with how Clements computes it in the paper).

File: strawberryfields/circuitspecs/chip0.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/circuitspecs/chip0.py]: Unitary decomposition always produces a diagonal array of local phase shifts. These can go anywhere, as they can be commuted through the beamsplitter arrays.

The `rectangular` function sandwiches them between two beamsplitter arrays, whereas `rectangular_symmetric` and `rectangular_phase_end` commute them to the end of the beamsplitter array --- this is preferred in experimental linear optics (https://github.com/XanaduAI/strawberryfields/pull/44).

So they are unavoidable, but having them at the end means they shouldn't affect the photon number statistics, and they can probably be ignored in hardware (@lneuhaus, did I get this right?)

File: strawberryfields/circuitspecs/chip0.py [hunk 14]
```diff
@@ -51,7 +59,130 @@ class Chip0Specs(CircuitSpecs):
...
+            return seq
+
+        # first do general GBS compilation to make sure
+        # Fock measurements are correct
+        # ---------------------------------------------
+        seq = GBSSpecs().compile(seq, registers)
+        A, B, C = group_operations(seq, lambda x: isinstance(x, ops.MeasureFock))
+
+        if len(B[0].reg) != self.modes:
+            raise CircuitError('All modes must be measured.')
```
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/circuitspecs/chip0.py]: Potentially, but does this make sense with hardware? It would mean throwing out results before returning it to the user.

With most qubit frameworks, all modes are assumed to terminate with measurement, even if the user doesn't specifically place a measurement operation (I like this approach). Some (Q#) even force all modes to be measured before allowing the code to compile (potentially too strict).

File: strawberryfields/circuitspecs/chip0.py [hunk 15]
```diff
@@ -51,7 +59,130 @@ class Chip0Specs(CircuitSpecs):
...
+            Command(ops.Interferometer(U01), registers[:2]),
+            Command(ops.Interferometer(U23), registers[2:])
+        ]
+
+        # decompose the interferometer, using Mach-Zehnder interferometers
+        B = self.decompose(B)
+
+        # Do a final circuit topology check
+        # ---------------------------------
+        seq = super().compile(A + B + C, registers)
```
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/circuitspecs/chip0.py]: :+1:

File: strawberryfields/circuitspecs/chip0.py [hunk 16]
```diff
@@ -51,7 +59,130 @@ class Chip0Specs(CircuitSpecs):
...
+        # -------------------------------------------------------------
+        A, B, C = group_operations(seq, lambda x: isinstance(x, (ops.Rgate, ops.BSgate)))
+
+        # begin unitary lists for mode [0, 1] and modes [2, 3] with
+        # two identity matrices. This is because multi_dot requires
+        # at least two matrices in the list.
+        U_list01 = [np.identity(self.modes//2, dtype=np.complex128)]*2
+        U_list23 = [np.identity(self.modes//2, dtype=np.complex128)]*2
+
+        if not C:
```
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/circuitspecs/chip0.py]: I originally had this check, but I _think_ it led to some weird edge case that I can't quite recall, and I had to change it.

I'll change this and double check that the tests pass.

File: strawberryfields/circuitspecs/circuit_specs.py [hunk 17]
```diff
@@ -212,3 +214,67 @@ def node_match(n1, n2):
...
+        should be applied is as follows.
+
+        1. First, we check if the operation is in :attr:`~.CircuitSpecs.decompositions`.
+           If not, decomposition is skipped, and the operation is applied
+             as a primitive (unless not supported by the ``CircuitSpec``).
+
+        2. Next, we check if the operation supports decomposition, and if the user
+           has explicitly requested no decomposition.
+
+           - If ``True``, the operation is applied
```
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/circuitspecs/circuit_specs.py]:
```suggestion
           - If both the above are true, the operation is applied
```

File: strawberryfields/ops.py [hunk 18]
```diff
@@ -1615,46 +1641,120 @@ def __or__(self, reg):
...
+      .. figure:: ../../_static/clements.png
+          :align: center
+          :width: 30%
+          :target: javascript:void(0);
+
+
+    * ``mesh='triangular'``: uses the scheme described in :cite:`reck1994`,
+      resulting in a *triangular* array of :math:`M(M-1)/2` beamsplitters arranged in
+      :math:`2M-3` slices and ordered from left to right and top to bottom. The
+      first and fourth beamsplitters act on wires :math:`M-1` and :math:`M`, the second
```
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/ops.py]: I copied this directly from https://pennylane.readthedocs.io/en/latest/code/templates/layers.html#pennylane.templates.layers.Interferometer :laughing: 

It's definitely hard to understand without looking at the corresponding diagram.

File: strawberryfields/ops.py [hunk 19]
```diff
@@ -1615,46 +1641,120 @@ def __or__(self, reg):
...
-            self.identity = True
-        else:
-            self.identity = False
-            self.BS1, self.BS2, self.R = clements(U, tol=tol)
+        allowed_meshes = {"rectangular", "rectangular_phase_end", "rectangular_symmetric", "triangular"}
+
+        if mesh not in allowed_meshes:
+            raise ValueError("Unknown mesh '{}'. Mesh must be one of {}".format(mesh, allowed_meshes))
+
+        self.identity = np.all(np.abs(U - np.identity(len(U))) < _decomposition_merge_tol)
```
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/ops.py]: Yep. I think there are still quite a few places in the code that uses this pattern - we should do a PR that changes all vestigal `np.all(np.abs(..) < ..)` to use `np.allclose`, much easier to read and understand.

File: tests/frontend/test_circuitspecs_chip0.py (hunk 4 above)
Comment (smite) [reply tests/frontend/test_circuitspecs_chip0.py]: The permutation probably happens during `lexicographical_topological_sort()`, since all measurement permutations are equivalent. The easiest fix is to add a `sorted()` call to `GBSSpec.compile()` using the subsystem index as the sorting key.

File: strawberryfields/ops.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (co9olguy) [reply strawberryfields/ops.py]: I looked at the diagram and still couldn't understand. Maybe we should update those diagrams with labels indicating the ordering explcitly. For instance, what does "top-to-bottom" ordering mean for the triangular mesh?

[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/ops.py]: Maybe I should just remove the text description, and just have the diagrams updated with labels?

File: tests/frontend/test_circuitspecs_chip0.py [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+        prog = sf.Program(4)
+        U = random_interferometer(2)
+
+        with prog.context as q:
+            ops.S2gate(0.5) | (q[0], q[2])
+            ops.Interferometer(U) | (q[0], q[1])
+            ops.Interferometer(U) | (q[2], q[3])
+            ops.MeasureFock() | q
+
+        with pytest.raises(CircuitError, match="S2gates placed on the incorrect modes"):
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/frontend/test_circuitspecs_chip0.py]: > It would be more clear if the error message said something like "S2gates do not appear on the correct modes"

Will make this change.

> 2nd question: does the particular mode pair matter, so long as it is consistent with the rest of the circuit topologically?

This is the reason why I am not doing much manual checking for the S2gates, and instead I'm just letting the final topology check catch any errors.

I was actually a bit torn with what to do here - we could go even further, and check which S2gates are correct (for more descriptive error messages), and add in identity S2gates if none are provided. But I decided this might be over-complicating it --- including the S2gates in the right place is much easier to do manually than the interferometer.

File: strawberryfields/circuitspecs/chip0.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/circuitspecs/chip0.py]: Just had a look, and the error messages look okay to me (for now).

The exceptions that could be raised from `GBSSpec.compile()`:

* `CircuitError: Operations following the Fock measurements.`

* `CircuitError: GBS circuits must contain Fock measurements.` (reads fine, since chip0 is a GBS device)

* `CircuitError: The Fock measurements are not consecutive.`

* `CircuitError: Measuring the same mode more than once.`

The exceptions that could be raised from `CircuitSpec.compile()`:

* `CircuitError: Program cannot be used with the CircuitSpec 'chip0' to incompatible topology.` - thoughts?

File: strawberryfields/ops.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/ops.py]: Done.

File: tests/frontend/test_circuitspecs_chip0.py [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+            ops.Rgate(0) | q[2]
+            ops.BSgate(np.pi/4, np.pi/2) | (q[2], q[3])
+            ops.Rgate(np.pi) | q[2]
+            ops.BSgate(np.pi/4, np.pi/2) | (q[2], q[3])
+            ops.Rgate(np.pi) | q[2]
+            ops.Rgate(0) | q[3]
+
+            ops.MeasureFock() | (q[0], q[3], q[1], q[2])
+
+        for cmd1, cmd2 in zip(res.circuit, expected.circuit):
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/frontend/test_circuitspecs_chip0.py]: Done (see `program_equivalence` function).

File: strawberryfields/ops.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/ops.py]: I've updated the text description to make it clearer and more concise.

File: strawberryfields/circuitspecs/gaussian.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (lneuhaus) [reply strawberryfields/circuitspecs/gaussian.py]: It does look nice to have all those gates with the same number of letters. If that is not a good enough reason, I do not object MZInterferometer.

File: tests/frontend/test_circuitspecs_chip0.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lneuhaus) [reply tests/frontend/test_circuitspecs_chip0.py]: `[0, 2*pi)` is the best for me.

File: strawberryfields/circuitspecs/chip0.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lneuhaus) [reply strawberryfields/circuitspecs/chip0.py]: Josh and I concluded this was the best way to modify the circuit spec to map an arbitrary unitary to the chip topology: if we do Fock measurements, this final phase is just a delay line before the photon number resolving detector, so it does not affect the measurement outcome. And in case one would perform Homodyne measurements one day, this final phase can be used comfortably to tweak the relative phase w.r.t. local oscillator (whose initial phase offset will probably be random). 

Only two minor comments:
- as discussed with Josh, another option could be to absorb an arbitrary phase shift into the FockMeasurement gate. This is just a suggestion based on the observation that any phase shift before a Fock Measurement does not affect the measurement outcome, and if it makes you feel uncomfortable, do not do it. 
- to keep names in line with `internal_phase_x` and `external_phase_y`, I suggest to call these final phases `final_phase_z` as this is more explicit than `phi_z`.

File: strawberryfields/circuitspecs/chip0.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (lneuhaus) [reply strawberryfields/circuitspecs/chip0.py]: In the near term, the hardware detector will perform a a measurement anyways. Not collecting the measurement result from a particular detection channel would of course reduce the network traffic and result memory footprint, so if there is enough time to implement this feature, it is not pointless (even though I do not see the use-case). In the more distant future, I can come up with two ideas for this:
- If detection channels should be a limiting factor some day (opposed to available chips), and an optical multiplexer would distribute the chip output modes among available detection channels, one could actually benefit from this as the output from more chips could be measured simultaneously. However, adding a multiplexer introduces loss and I therefore doubt this will ever be a useful scenario, as opposed to having a fixed (spliced) routing from chip output modes to detector channels.
- Maybe this makes sense if a circuit is expected to be "plugged" into another circuit one day?

File: strawberryfields/ops.py [hunk 22]
```diff
@@ -1615,46 +1645,123 @@ def __or__(self, reg):
...
+
+            - ``'rectangular'`` - rectangular mesh, with local phase shifts
+              applied between interferometers
+
+            - ``'rectangular_phase_end'`` - rectangular mesh, with local phase shifts
+              placed after all interferometers
+
+            - ``'rectangular_symmetric'`` - rectangular mesh, with local phase shifts
+              placed after all interferometers, and all beamsplitters decomposed into
+              symmetric beamsplitters
```
Comment (lneuhaus) [inline strawberryfields/ops.py]:
```suggestion
              pairs of symmetric beamsplitters and phase shifters
```

Comment (lneuhaus) [reply strawberryfields/ops.py]: I did not look at all lines of code in detail. If you would like me to do that, let me know and I will do that.

File: strawberryfields/circuitspecs/chip0.py [hunk 23]
```diff
@@ -51,7 +62,134 @@ class Chip0Specs(CircuitSpecs):
...
+            raise CircuitError(
+                "Interferometer on modes [0, 1] must be identical to interferometer on modes [2, 3]."
+            )
+
+        U = block_diag(U01, U23)
+
+        # replace B with an interferometer
+        B = [
+            Command(ops.Interferometer(U01), registers[:2]),
+            Command(ops.Interferometer(U23), registers[2:]),
```
Comment (smite) [inline strawberryfields/circuitspecs/chip0.py]: Is this the only place where `registers` is needed? Would it be possible to deduce the correct RegRefs by looking at the `Command` instances in `seq`?

Comment (josh146) [reply strawberryfields/circuitspecs/chip0.py]: I ran into an issue where a circuit with no unitary (which should compile fine), I could not apply an identity `Interferometer` decomposition because I did not have access to the `regref` instances. While I could have accessed the regref instances from the `MeasureFock` commands, this felt hacky, and there may be more use cases in the future (e.g., compiling an empty program).

File: strawberryfields/circuitspecs/chip0.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/circuitspecs/chip0.py]: > to keep names in line with internal_phase_x and external_phase_y, I suggest to call these final phases final_phase_z as this is more explicit than phi_z.

done :+1:

File: strawberryfields/circuitspecs/chip0.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/circuitspecs/chip0.py]: Let's make this an issue, as at the moment SF does not have the capability to specify a topology with optional gates

File: tests/frontend/test_circuitspecs_chip0.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (josh146) [reply tests/frontend/test_circuitspecs_chip0.py]: Fixed, now `Interferometer(U, mesh="rectangular_symmetric")` now always enforces `phi % (2*np.pi)` on all phase angles.

**Comment by josh146** [comment]
I think this is ready for re-review.

File: tests/frontend/test_circuitspecs_chip0.py [hunk 24]
```diff
@@ -0,0 +1,379 @@
...
+        p_match = np.allclose(n1["p"], n2["p"], atol=atol, rtol=rtol)
+        wire_match = n1["w"] == n2["w"]
+
+        if compare_params:
+            return name_match and p_match and wire_match
+
+        return name_match and wire_match
+
+    # check if circuits are equivalent
+    return nx.is_isomorphic(circuit[0], circuit[1], node_match)
```
Comment (josh146) [inline tests/frontend/test_circuitspecs_chip0.py]: Note: we could move this into `program_utils` in a future PR to satisfy #135.

File: strawberryfields/circuitspecs/chip0.py (hunk 3 above)
Comment (smite) [reply strawberryfields/circuitspecs/chip0.py]: I think you can call a non-abstract method of an ABC, `ABC.methodname(self_param, other_params)`

Comment (josh146) [reply strawberryfields/circuitspecs/chip0.py]: I tried this, I got `TypeError: Can't instantiate abstract class CircuitSpecs with abstract methods`

**Comment by co9olguy** [comment]
Does the PR yet include S2gates?

**Comment by josh146** [comment]
> Does the PR yet include S2gates?

No, I thought it was out of scope for this PR. This PR simply provides the scaffolding for interferometer compilation - once merged, it can be easily re-used for any other device that requires compilation of a unitary.

I have a separate branch (https://github.com/XanaduAI/strawberryfields/tree/bipartite_graph_embed) containing a `BipartiteGraphEmbed`, that, for a bipartite graph with degenerate singular values, runs `GraphEmbed` and converts the `Sgate`s to `S2gate`s.

File: tests/frontend/test_circuitspecs_chip0.py [hunk 25]
```diff
@@ -317,17 +320,17 @@ def test_mach_zehnder_interferometers(self, tol):
             ops.S2gate(0.5, 0) | (q[1], q[3])
 
             # corresponds to MZgate(phi, theta) on modes [0, 1]
-            ops.Rgate(phi) | q[0]
+            ops.Rgate(phi % (2*np.pi)) | q[0]
```
Comment (co9olguy) [inline tests/frontend/test_circuitspecs_chip0.py]: how come you used `np.mod` elsewhere?

File: tests/frontend/test_ops_decompositions.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (co9olguy) [reply tests/frontend/test_ops_decompositions.py]: might as well fix here, it's a one-liner to delete them

**Comment by co9olguy** [comment]
Approving, conditional on one thing:
There were a number of questions/issues/suggestions that came up during code review. Before merging, can you please document the unresolved comments above with new issues? I don't want to forget these things. Also make an issue to add S2gates to chip0 compilalation :laughing:

File: tests/frontend/test_circuitspecs_chip0.py (hunk 25 above)
Comment (josh146) [reply tests/frontend/test_circuitspecs_chip0.py]: Oh. Just me not being consistent. I can change everything over to `np.mod`.

File: strawberryfields/ops.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/ops.py]: Fixed.

File: tests/frontend/test_circuitspecs_chip0.py (hunk 25 above)
Comment (josh146) [reply tests/frontend/test_circuitspecs_chip0.py]: Done

File: tests/frontend/test_ops_decompositions.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (josh146) [reply tests/frontend/test_ops_decompositions.py]: done
